### PR TITLE
add bound vars to op_argument_list

### PR DESCRIPTION
**Pull request by AviAvni**
fix https://github.com/FalkorDB/FalkorDB/issues/699

## Summary by CodeRabbit

- **New Features**
	- Enhanced the `FOREACH` loop in Cypher queries to support additional bound variables, improving flexibility in graph database operations.
- **Tests**
	- Added a new test to validate the behavior of bound variables within a `FOREACH` loop in Cypher queries.

### CODE REVIEW

File: src/execution_plan/execution_plan_build/execution_plan_construct.c [hunk 1]
```diff
@@ -233,6 +233,7 @@ static void _buildForeachOp
 	ExecutionPlan *embedded_plan = ExecutionPlan_NewEmptyExecutionPlan();
 	embedded_plan->ast_segment = body_ast;
 	embedded_plan->record_map = raxClone(plan->record_map);
+	const char **arguments = (const char **)raxKeys(embedded_plan->record_map);
```
Comment (swilly22) [inline src/execution_plan/execution_plan_build/execution_plan_construct.c:236]: Dangerous, for the reason where `record_map` might include variables which are bounded after the `FOREACH` clause. Moreover (just making sure) the `arguments` list should include the loop variable e.g. `FOREACH p IN posts` should `p` be in `arguments`?

File: src/execution_plan/ops/op_argument_list.c [hunk 2]
```diff
@@ -82,7 +88,7 @@ static inline OpBase *ArgumentListClone
 	const OpBase *opBase
 ) {
 	ASSERT(opBase->type == OPType_ARGUMENT_LIST);
-	return NewArgumentListOp(plan);
+	return NewArgumentListOp(plan, opBase->modifies);
```
Comment (swilly22) [inline src/execution_plan/ops/op_argument_list.c:91]: Does `opBase->modifies` is of type `arr.h` ?

File: src/execution_plan/ops/op_argument_list.h [hunk 3]
```diff
@@ -20,7 +20,8 @@ typedef struct {
 
 OpBase *NewArgumentListOp
 (
-	const ExecutionPlan *plan
+	const ExecutionPlan *plan,
+	const char **variables
```
Comment (swilly22) [inline src/execution_plan/ops/op_argument_list.h:24]: Consider renaming to `bound_vars`
would be nice to add a bit of comments / documentation

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/execution_plan/execution_plan_build/execution_plan_construct.c (hunk 1 above)
Comment (AviAvni) [reply src/execution_plan/execution_plan_build/execution_plan_construct.c:236]: in this location record_map contains only the variables befor the foreach is constructed
p is in the unwind modified

File: src/execution_plan/ops/op_argument_list.c (hunk 2 above)
Comment (AviAvni) [reply src/execution_plan/ops/op_argument_list.c:91]: yes took the same code from op_argument

File: src/execution_plan/ops/op_argument_list.h (hunk 3 above)
Comment (AviAvni) [reply src/execution_plan/ops/op_argument_list.h:24]: I took the same code from op_argument

File: src/execution_plan/execution_plan_build/execution_plan_construct.c (hunk 1 above)
Comment (swilly22) [reply src/execution_plan/execution_plan_build/execution_plan_construct.c:236]: I think this is a bit dangerous to always assume, for safety reasons can we switch to do bound entities collection via execution-plan traversal?

File: src/execution_plan/execution_plan_build/execution_plan_construct.c [hunk 4]
```diff
@@ -233,6 +233,13 @@ static void _buildForeachOp
 	ExecutionPlan *embedded_plan = ExecutionPlan_NewEmptyExecutionPlan();
 	embedded_plan->ast_segment = body_ast;
 	embedded_plan->record_map = raxClone(plan->record_map);
+	const char **arguments = NULL;
+	if(plan->root) {
+		rax *bound_vars = raxNew();
+		ExecutionPlan_BoundVariables(plan->root, bound_vars, plan);
+		arguments = (const char **)raxValues(bound_vars);
+		raxFree(bound_vars);
+	}
```
Comment (swilly22) [inline src/execution_plan/execution_plan_build/execution_plan_construct.c:242]: collect bound variables from the current operation, not the plan's root.

**Review by swilly22 — CHANGES_REQUESTED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**